ADBDEV-3674: ORCA initialization refactoring

RekGRpth opened the pull request:
ORCA initialization refactoring

When initializing ([CGPOptimizer::InitGPOPT](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gpopt/CGPOptimizer.cpp#L171-L184)) in the [gpos_init](https://github.com/arenadata/gpdb/blob/88532d712d0258ab14e5992854bd04605d1e763f/src/backend/gporca/libgpos/src/_api.cpp#L129-L169) function, ORCA mixed a C approach (it returned error codes and then checked them) and a C++ approach (it threw exceptions and then caught them).

1\. In case of low memory or any other exception, the [CWorkerPoolManager::Init](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/task/CWorkerPoolManager.cpp#L83-L88) function did not return GPOS_OK, and then the [gpos_init](https://github.com/arenadata/gpdb/blob/88532d712d0258ab14e5992854bd04605d1e763f/src/backend/gporca/libgpos/src/_api.cpp#L141) function called the [CMemoryPoolManager::GetMemoryPoolMgr()->Shutdown](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/memory/CMemoryPoolManager.cpp#L274) function, which reset CMemoryPoolManager::m_memory_pool_mgr.
In this case, all the exceptions that arose were not rethrown, which means that the execution continued, and the following [gpdxl_init]((https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libnaucrates/src/init.cpp#L126)) function was called, in which the [CAutoMemoryPool](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/memory/CAutoMemoryPool.cpp#L42-L46) class was initialized.
In the constructor of this class, the static method [CMemoryPoolManager::GetMemoryPoolMgr](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/include/gpos/memory/CMemoryPoolManager.h#L194-L198) was called, which returned NULL, because we nullified m_memory_pool_mgr up the stack. Thus, a segfault could occur.

We can emulate OOM by setting the block variable to zero on this stack
```gdb
AllocSetAllocImpl(MemoryContext context, Size size, bool isHeader) (aset.c:1377)
AllocSetAlloc(MemoryContext context, Size size) (aset.c:1454)
MemoryContextAlloc(MemoryContext context, Size size) (mcxt.c:1196)
gpdb::GPDBMemoryContextAlloc(MemoryContext context, Size size) (gpdbwrappers.cpp:2694)
gpos::CMemoryPoolPalloc::NewImpl(gpos::CMemoryPoolPalloc * const this, const gpos::ULONG bytes, gpos::CMemoryPool::EAllocationType eat) (CMemoryPoolPalloc.cpp:40)
operator new(gpos::SIZE_T size, gpos::CMemoryPool * mp, const gpos::CHAR * filename, gpos::ULONG line) (CMemoryPool.h:280)
gpos::CWorkerPoolManager::Init() (CWorkerPoolManager.cpp:74)
gpos_init(gpos_init_params * params) (_api.cpp:139)
CGPOptimizer::InitGPOPT() (CGPOptimizer.cpp:181)
InitGPOPT() (CGPOptimizer.cpp:248)
InitPostgres(const char * in_dbname, Oid dboid, const char * username, char * out_dbname) (postinit.c:664)
PostgresMain(int argc, char ** argv, const char * dbname, const char * username) (postgres.c:4898)
BackendRun(Port * port) (postmaster.c:4831)
BackendStartup(Port * port) (postmaster.c:4495)
ServerLoop() (postmaster.c:1956)
PostmasterMain(int argc, char ** argv) (postmaster.c:1526)
main(int argc, char ** argv) (main.c:245)
```
Or we can emulate it more simply by explicitly throwing an exception
```diff
diff --git a/src/backend/gporca/libgpos/src/task/CWorkerPoolManager.cpp b/src/backend/gporca/libgpos/src/task/CWorkerPoolManager.cpp
index 7f98380d3e..7a40aa0908 100644
--- a/src/backend/gporca/libgpos/src/task/CWorkerPoolManager.cpp
+++ b/src/backend/gporca/libgpos/src/task/CWorkerPoolManager.cpp
@@ -69,6 +69,7 @@ CWorkerPoolManager::Init()
 
        GPOS_TRY
        {
+               GPOS_ASSERT(!"fail");
                // create worker pool
                CWorkerPoolManager::m_worker_pool_manager =
                        GPOS_NEW(mp) CWorkerPoolManager(mp);
```
```gdb
#0  raise () from /lib64/libpthread.so.0
#1  StandardHandlerForSigillSigsegvSigbus_OnMainThread (processName=0x14ee852 "Process", postgres_signal_arg=11) at elog.c:5529
#2  CdbProgramErrorHandler (postgres_signal_arg=11) at postgres.c:3587
#3  <signal handler called>
#4  gpos::CMemoryPoolManager::CreateMemoryPool (this=0x0) at CMemoryPoolManager.cpp:85
#5  gpos::CAutoMemoryPool::CAutoMemoryPool (this=0x7ffc149f6080, leak_check_type=gpos::CAutoMemoryPool::ElcExc) at CAutoMemoryPool.cpp:45
#6  gpdxl_init () at init.cpp:126
#7  CGPOptimizer::InitGPOPT () at CGPOptimizer.cpp:182
#8  InitGPOPT () at CGPOptimizer.cpp:248
#9  InitPostgres (in_dbname=0x8b915d0 "postgres", dboid=0, username=0x8b54be8 "postgres", out_dbname=0x0) at postinit.c:662
#10 PostgresMain (argc=1, argv=0x8b91630, dbname=0x8b915d0 "postgres", username=0x8b54be8 "postgres") at postgres.c:4898
#11 BackendRun (port=0x8b7be00) at postmaster.c:4831
#12 BackendStartup (port=0x8b7be00) at postmaster.c:4495
#13 ServerLoop () at postmaster.c:1956
#14 PostmasterMain (argc=7, argv=0x8b52940) at postmaster.c:1526
#15 main (argc=7, argv=0x8b52940) at main.c:245
```

2\. If during the initialization of ORCA an exception occurred that was not caught by the initialization functions, then it was caught in the [InitGPOPT](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gpopt/CGPOptimizer.cpp#L242-L259) function
and was silently ignored, simply disabling ORCA. Further, when the process terminated (even successfully), the [gpopt_terminate](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gpopt/CGPOptimizer.cpp#L197) function was called.
On a debug build, it tried to free the [CMemoryPoolManager::GetMemoryPoolMgr()->Destroy(mp)](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpopt/src/init.cpp#L75) memory pool.
But due to an exception during ORCA initialization, this memory pool was not created and therefore assert [GPOS_ASSERT(NULL != mp)](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/memory/CMemoryPoolManager.cpp#L112) fired.
In a debug build, assert turned into an exception that was caught and silently ignored in the [TerminateGPOPT](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gpopt/CGPOptimizer.cpp#L269-L286) function.
In this case, the Shutdown method of the CWorkerPoolManager class was not called, in which the [CWorkerPoolManager::m_worker_pool_manager](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/task/CWorkerPoolManager.cpp#L122) variable would be set to zero.
When deleting the context [MemoryContextDelete(OptimizerMemoryContext)](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/utils/init/postinit.c#L1413) on a debug build, all freed memory was filled with the sequence 7f,
including the m_single_worker variable of the m_worker_pool_manager instance of the CWorkerPoolManager class.
At the very end of the process, the destructor CGroupExpression::~CGroupExpression of the global static variable [CGroupExpression::m_gexprInvalid](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpopt/src/search/CGroupExpression.cpp#L38) was called.
And then the destructor of its parent class CRefCount::~CRefCount, in which (again, on a debug build) a sequence of methods was called
[ITask::Self](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/include/gpos/common/CRefCount.h#L68)
[IWorker *worker = IWorker::Self](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/task/ITask.cpp#LL29C2-L29C36)
[worker = CWorkerPoolManager::WorkerPoolManager()->Self](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/task/IWorker.cpp#LL35C3-L35C60)
which resulted in worker = 0x7f7f7f7f7f7f7f7f and a segfault occurred when calling [return worker->GetTask](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/task/ITask.cpp#LL32C3-L32C27).

We can emulate OOM by setting the block variable to zero on this stack
```gdb
AllocSetAllocImpl(MemoryContext context, Size size, bool isHeader) (aset.c:1377)
AllocSetAlloc(MemoryContext context, Size size) (aset.c:1454)
MemoryContextAlloc(MemoryContext context, Size size) (mcxt.c:1196)
gpdb::GPDBMemoryContextAlloc(MemoryContext context, Size size) (gpdbwrappers.cpp:2694)
gpos::CMemoryPoolPalloc::NewImpl(gpos::CMemoryPoolPalloc * const this, const gpos::ULONG bytes, gpos::CMemoryPool::EAllocationType eat) (CMemoryPoolPalloc.cpp:40)
operator new(gpos::SIZE_T size, gpos::CMemoryPool * mp, const gpos::CHAR * filename, gpos::ULONG line) (CMemoryPool.h:280)
gpopt::CXformFactory::Init() (CXformFactory.cpp:369)
gpopt_init() (init.cpp:55)
CGPOptimizer::InitGPOPT() (CGPOptimizer.cpp:183)
InitGPOPT() (CGPOptimizer.cpp:248)
InitPostgres(const char * in_dbname, Oid dboid, const char * username, char * out_dbname) (postinit.c:664)
PostgresMain(int argc, char ** argv, const char * dbname, const char * username) (postgres.c:4898)
BackendRun(Port * port) (postmaster.c:4831)
BackendStartup(Port * port) (postmaster.c:4495)
ServerLoop() (postmaster.c:1956)
PostmasterMain(int argc, char ** argv) (postmaster.c:1526)
main(int argc, char ** argv) (main.c:245)
```
Or we can emulate it more simply by explicitly throwing an exception
```diff
diff --git a/src/backend/gporca/libgpopt/src/xforms/CXformFactory.cpp b/src/backend/gporca/libgpopt/src/xforms/CXformFactory.cpp
index dcfd0e2187..578f2b13f6 100644
--- a/src/backend/gporca/libgpopt/src/xforms/CXformFactory.cpp
+++ b/src/backend/gporca/libgpopt/src/xforms/CXformFactory.cpp
@@ -362,6 +362,7 @@ CXformFactory::Init()
        CMemoryPool *mp = CMemoryPoolManager::CreateMemoryPool();
        GPOS_TRY
        {
+               GPOS_ASSERT(!"fail");
                // create xform factory instance
                m_pxff = GPOS_NEW(mp) CXformFactory(mp);
        }
```
```gdb
#0  raise (sig=11) at ../nptl/sysdeps/unix/sysv/linux/pt-raise.c:36
#1  StandardHandlerForSigillSigsegvSigbus_OnMainThread (processName=0x14ee852 "Process", postgres_signal_arg=11) at elog.c:5529
#2  CdbProgramErrorHandler (postgres_signal_arg=11) at postgres.c:3587
#3  <signal handler called>
#4  gpos::ITask::Self () at ITask.cpp:32
#5  gpos::CRefCount::~CRefCount (this=0x775fda0 <gpopt::CGroupExpression::m_gexprInvalid>, __in_chrg=<optimized out>) at ../../../../../../src/backend/gporca/libgpos/include/gpos/common/CRefCount.h:68
#6  gpopt::CGroupExpression::~CGroupExpression (this=0x775fda0 <gpopt::CGroupExpression::m_gexprInvalid>, __in_chrg=<optimized out>) at CGroupExpression.cpp:112
#7  __run_exit_handlers (status=0, listp=0x7f243d5c76c8 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true) at exit.c:77
#8  __GI_exit (status=<optimized out>) at exit.c:99
#9  proc_exit (code=0) at ipc.c:145
#10 PostgresMain (argc=1, argv=0x8e079f0, dbname=0x8e07990 "postgres", username=0x8e07968 "postgres") at postgres.c:5764
#11 BackendRun (port=0x8e2efb0) at postmaster.c:4831
#12 BackendStartup (port=0x8e2efb0) at postmaster.c:4495
#13 ServerLoop () at postmaster.c:1956
#14 PostmasterMain (argc=6, argv=0x8e05950) at postmaster.c:1526
#15 main (argc=6, argv=0x8e05950) at main.c:245
```

3\. A single instance of the CWorker class was created only in the [gpos_exec](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/_api.cpp#L207) function, which was called not when the backend was initialized, but much later - when planning the request.
At the same time, the CWorker::CWorker constructor [registered](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/task/CWorker.cpp#L43) it in an instance of the CWorkerPoolManager class.
Thus, when initializing ORCA, the CWorkerPoolManager::WorkerPoolManager()->Self method returned NULL.
If an exception occurred when using the CAutoMemoryPool class allocated on the stack, for example, there was not enough memory, then the CAutoMemoryPool::~CAutoMemoryPool destructor was still called.
On the debug build, a sequence of methods was called in it
[ITask *task = ITask::Self](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/memory/CAutoMemoryPool.cpp#L91)
[IWorker *worker = IWorker::Self](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/task/ITask.cpp#LL29C2-L29C36)
[worker = CWorkerPoolManager::WorkerPoolManager()->Self](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/task/IWorker.cpp#LL35C3-L35C60)which, when initializing ORCA, led to task = NULL and if ElcExc == m_leak_check_type, then assert [GPOS_ASSERT_IMP(ElcExc == m_leak_check_type, NULL != task)](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/memory/CAutoMemoryPool.cpp#LL94C18-L94C45) was triggered.
On the debug build, the assert turned into an exception, and now two exceptions became active in the destructor at the same time, which led to a crash.

We can emulate OOM by setting the block variable to zero on this stack
```gdb
AllocSetAllocImpl(MemoryContext context, Size size, bool isHeader) (aset.c:1377)
AllocSetAlloc(MemoryContext context, Size size) (aset.c:1454)
MemoryContextAlloc(MemoryContext context, Size size) (mcxt.c:1196)
gpdb::GPDBMemoryContextAlloc(MemoryContext context, Size size) (gpdbwrappers.cpp:2694)
gpos::CMemoryPoolPalloc::NewImpl(gpos::CMemoryPoolPalloc * const this, const gpos::ULONG bytes, gpos::CMemoryPool::EAllocationType eat) (CMemoryPoolPalloc.cpp:40)
operator new(gpos::SIZE_T size, gpos::CMemoryPool * mp, const gpos::CHAR * filename, gpos::ULONG line) (CMemoryPool.h:280)
gpos::CMessageRepository::Init() (CMessageRepository.cpp:114)
gpos_init(gpos_init_params * params) (_api.cpp:136)
CGPOptimizer::InitGPOPT() (CGPOptimizer.cpp:181)
InitGPOPT() (CGPOptimizer.cpp:260)
InitPostgres(const char * in_dbname, Oid dboid, const char * username, char * out_dbname) (postinit.c:664)
PostgresMain(int argc, char ** argv, const char * dbname, const char * username) (postgres.c:4898)
BackendRun(Port * port) (postmaster.c:4831)
BackendStartup(Port * port) (postmaster.c:4495)
ServerLoop() (postmaster.c:1956)
PostmasterMain(int argc, char ** argv) (postmaster.c:1526)
main(int argc, char ** argv) (main.c:245)
```
Or we can emulate it more simply by explicitly throwing an exception
```diff
diff --git a/src/backend/gporca/libgpos/src/error/CMessageRepository.cpp b/src/backend/gporca/libgpos/src/error/CMessageRepository.cpp
index 5785eaea5a..ab8e3704be 100644
--- a/src/backend/gporca/libgpos/src/error/CMessageRepository.cpp
+++ b/src/backend/gporca/libgpos/src/error/CMessageRepository.cpp
@@ -115,6 +115,7 @@ CMessageRepository::Init()
        repository->InitDirectory(mp);
        repository->LoadStandardMessages();
 
+       GPOS_ASSERT(!"fail");
        CMessageRepository::m_repository = repository;
 
        // detach safety
```
```gdb
#0  __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:55
#1  __GI_abort () at abort.c:90
#2  __gnu_cxx::__verbose_terminate_handler () at ../../../../libstdc++-v3/libsupc++/vterminate.cc:95
#3  __cxxabiv1::__terminate (handler=<optimized out>) at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:38
#4  __cxa_call_terminate (ue_header=0x9438b50) at ../../../../libstdc++-v3/libsupc++/eh_call.cc:54
#5  __cxxabiv1::__gxx_personality_v0 (version=<optimized out>, actions=<optimized out>, exception_class=<optimized out>, ue_header=<optimized out>, context=<optimized out>) at ../../../../libstdc++-v3/libsupc++/eh_personality.cc:676
#6  _Unwind_RaiseException_Phase2 (exc=exc@entry=0x9438b50, context=context@entry=0x7fff11b65940) at ../../../libgcc/unwind.inc:62
#7  _Unwind_Resume (exc=0x9438b50) at ../../../libgcc/unwind.inc:230
#8  gpos::CAutoMemoryPool::~CAutoMemoryPool (this=0x7fff11b65ae0, __in_chrg=<optimized out>) at CAutoMemoryPool.cpp:128
#9  gpos::CMessageRepository::Init () at CMessageRepository.cpp:125
#10 gpos_init (params=0x7fff11b65b48) at _api.cpp:145
#11 CGPOptimizer::InitGPOPT () at CGPOptimizer.cpp:181
#12 InitGPOPT () at CGPOptimizer.cpp:248
#13 InitPostgres (in_dbname=0x944cfe0 "postgres", dboid=0, username=0x9410be8 "postgres", out_dbname=0x0) at postinit.c:662
#14 PostgresMain (argc=1, argv=0x944d040, dbname=0x944cfe0 "postgres", username=0x9410be8 "postgres") at postgres.c:4898
#15 BackendRun (port=0x9437e00) at postmaster.c:4831
#16 BackendStartup (port=0x9437e00) at postmaster.c:4495
#17 ServerLoop () at postmaster.c:1956
#18 PostmasterMain (argc=7, argv=0x940e940) at postmaster.c:1526
#19 main (argc=7, argv=0x940e940) at main.c:245
```

The patch refactors the ORCA initialization code, getting rid of the C-approach, i.e. removes the return of error codes and their checking, and adds a C++ approach, i.e. catching exceptions and rethrowing them.
It also causes the backend to terminate with error on any exception during ORCA initialization.
The gpdxl_init function has been changed, because the auto memory pool is used to automatically free memory when it goes out of scope, but here it is immediately removed from it, and in fact, it is not needed at all.

RekGRpth commented:
I'll update the commit description if that description is clear.

andr-sokolov commented on src/backend/gporca/libgpopt/include/gpopt/xforms/CXformFactory.h:
```diff
@@ -111,10 +111,10 @@ class CXformFactory
 
 	// destroy global factory instance
-	void Shutdown();
+	static void Shutdown();
```
This change does not depend on replacing of error codes checking with interception of exceptions. So adding "static" should be in another commit

RekGRpth replied:
No! See using CXformFactory::Shutdown.

RekGRpth replied:
done

andr-sokolov commented on src/backend/gporca/libgpopt/src/init.cpp:
```diff
@@ -42,20 +42,27 @@ void
 {
 	{
-		CAutoMemoryPool amp;
+		CAutoMemoryPool amp(CAutoMemoryPool::ElcStrict);
```
Why has CAutoMemoryPool::ElcExc been replaced with CAutoMemoryPool::ElcStrict?

RekGRpth replied:
> Why has CAutoMemoryPool::ElcExc been replaced with CAutoMemoryPool::ElcStrict?

To prevent the second scenario.

andr-sokolov commented on src/backend/gporca/libgpopt/src/xforms/CXformFactory.cpp:
```diff
@@ -405,18 +392,18 @@ CXformFactory::Init()
-	GPOS_ASSERT(NULL != pxff && "Xform factory has not been initialized");
-
-	CMemoryPool *mp = pxff->m_mp;
+	if (NULL != m_pxff)
```
I suggest to use
```
if (NULL == m_pxff)
    return;
```
because this way we can reduce indentation and number of modified lines

RekGRpth replied:
I did it in a similar way to https://github.com/arenadata/gpdb/blob/db2d3880fb989ef49afa121d13869f768fd2c901/src/backend/gporca/libgpos/src/common/CDebugCounter.cpp#L72-L83

andr-sokolov replied:
Why is it better than the variant which I suggest?

RekGRpth replied:
> Why is it better than the variant which I suggest?

It's not better, it's just that it's already been done this way in the specified place, and I did it by analogy as well.

RekGRpth replied:
done

andr-sokolov commented on src/backend/gporca/libgpos/src/error/CMessageRepository.cpp:
```diff
@@ -103,12 +103,12 @@ CMessageRepository::LookupMessage(CException exc, ELocale locale)
 	GPOS_ASSERT(NULL == m_repository);
 
-	CAutoMemoryPool amp;
+	CAutoMemoryPool amp(CAutoMemoryPool::ElcStrict);
```
This change fixes the second case and does not depend on other changes. So I suggest to place it in a separate commit.

RekGRpth replied:
As you can see, I added this argument to the constructor not only in this place, but also in many other places when initializing the ORCA, because a scenario similar to the second case can also occur in them.

andr-sokolov replied:
The constructor argument has been changed to avoid an exception in the destructor. It will be better to exclude raising exception in the destructor at all.

RekGRpth replied:
done

andr-sokolov commented:
My version of description for the second case:

When OOM happens in `MessageRepository::Init()`, for example during initialization of the `repository` variable, then an exception occurs, which cause to calling of the CAutoMemoryPool class destructor before calling of its Detach() method. Because the Detach() method is not called, the `NULL == m_mp` condition is false in the destructor, so the destructor execution continues and the task variable is set to NULL. This cause to the violation of the `GPOS_ASSERT_IMP(ElcExc == m_leak_check_type, NULL != task);` restriction. So a new exception is thrown at the time of processing another exception. As a result the program crashes without displaying an informative error message.
The `task` variable is set to NULL, because inside of `ITask::Self()` the `CWorkerPoolManager::WorkerPoolManager()->Self()` method is called, which returns NULL only at the time of ORCA initialization. The reason is that an instance of the CWorker class is created only in the gpos_exec function, which is called not when a backend is initialized, but much later - when planning a query. The CWorker class constructor calls the `CWorkerPoolManager::WorkerPoolManager()->RegisterWorker()` method. After that the `CWorkerPoolManager::WorkerPoolManager()->Self()` method returns not NULL value.

RekGRpth commented:
> When OOM happens in `MessageRepository::Init()`

You misunderstood! The case in `MessageRepository::Init()` is just one example of a bug with `CAutoMemoryPool` being allocated on the stack during `ORCA` initialization, when an exception occurs and the destructor is called, in which another exception occurs.

RekGRpth commented:
What means

> during initialization of the `repository` variable

?

RekGRpth commented:
> an exception occurs, which cause to calling of the CAutoMemoryPool class destructor

No! An exception does not cause the destructor to be called!

HustonMmmavr commented on src/backend/gporca/libgpos/src/_api.cpp:
```diff
@@ -131,41 +131,41 @@ gpos_init(struct gpos_init_params *params)
+		gpos::CMemoryPoolManager::Init();
+		gpos::CWorkerPoolManager::Init();
+		gpos::CMessageRepository::Init();
+		gpos::CCacheFactory::Init();
```
Does it is a fatal that cache factory wasn't created? Seems previous approach ignores it
And what about CDebugCounter::Init() - is it a fatal?

RekGRpth replied:
> Seems previous approach ignores it

The previous approach ignored a lot of things.

RekGRpth replied:
> Does it is a fatal that cache factory wasn't created?

standard_planner->optimize_query->GPOPTOptimizedPlan->CGPOptimizer::GPOPTOptimizedPlan->COptTasks::GPOPTOptimizedPlan->COptTasks::OptimizeTask->CMDCache::Init->CCacheFactory::CreateCache->CCacheFactory::GetFactory

RekGRpth replied:
> And what about CDebugCounter::Init() - is it a fatal?

CDebugCounter is only used with GPOS_DEBUG_COUNTERS.
So, if this macro is already defined, then the memory for the counter must also be successfully allocated.

HustonMmmavr commented on src/backend/gporca/libgpos/src/memory/CCacheFactory.cpp:
```diff
@@ -80,17 +78,9 @@ CCacheFactory::Init()
 
 		CCacheFactory::m_factory = NULL;
 
-		if (GPOS_MATCH_EX(ex, CException::ExmaSystem, CException::ExmiOOM))
```
Use the same approach as at CMessageRepository/CFSSimulator/CDebugCounters - without try/catch?

RekGRpth replied:
> Use the same approach as at CMessageRepository/CFSSimulator/CDebugCounters - without try/catch?

Yes, but then you need to change it to CAutoMemoryPool, and there is an error in its destructor, so it's better to pass it to the specified classes in order not to use CAutoMemoryPool in them.
In addition, such refactoring will increase the already rather big diff.

HustonMmmavr commented on src/backend/gporca/libgpos/src/error/CFSimulator.cpp:
```diff
@@ -158,7 +158,10 @@ CFSimulator::Init()
+	if (NULL == m_fsim)
+		return;
+
+	CMemoryPool *mp = m_fsim->m_mp;
```
Why not to do like (and shutdown of other classes, related to this patch, like at CMessageRepository)?
Why does we need to store the mp at separate variable and previously cleanup and null the m_fsim, and after that
delete mp? 
```cc
CMemoryPoolManager::GetMemoryPoolMgr()->Destroy(m_flsim->mp);
GPOS_DELETE(m_fsim);
m_fsim = NULL
```

HustonMmmavr replied:
If code from message above is not correct (the deletion of fsim should be performed firstly), seems the next block would be more readable:
```cc
CMemoryPool *mp = CFSimulator::m_fsim->mp;
GPOS_DELETE(CFSimulator::m_fsim);

CFSimulator::m_fsim = NULL;
CMemoryPoolManager::GetMemoryPoolMgr()->Destroy(mp);
```
And make the same approach at other shutdown methods

RekGRpth replied:
> Why not to do like (and shutdown of other classes, related to this patch, like at CMessageRepository)?

Here it is done to reduce the diff.

RekGRpth replied:
> Why does we need to store the mp at separate variable and previously cleanup and null the m_fsim, and after that
> delete mp?

GPOS_DELETE uses a memory pool.

RekGRpth replied:
> seems the next block would be more readable

Just wrapping an empty line will increase the diff again.

HustonMmmavr replied:
if this code is valid (from the first message):
```
CMemoryPoolManager::GetMemoryPoolMgr()->Destroy(m_flsim->mp);
GPOS_DELETE(m_fsim);
m_fsim = NULL
```

Seems it's worth to refactor (at this and other classes). Yes it will increase the diff, but what is the purpose of reducing the diff? It may be more complex to push a refactor changes to upstream, and in addition this the single approach, used at Shutdown methods is easier to review

HustonMmmavr commented on src/backend/gporca/libgpos/src/memory/CCacheFactory.cpp:
```diff
@@ -105,10 +95,10 @@ CCacheFactory::Init()
+	if (NULL == m_factory)
+		return;
 
+	CCacheFactory *factory = CCacheFactory::GetFactory();
```
Does we really need to use a getter here (`GetFactory`)? For me - it's a bit confusing, above `if` condition checks m_factory, but here the used a getter of this variable, and code bellow nulls the variable  itself `CCacheFactory::m_factory = NULL;`. Maybe it's worth to unify Shutdown approaches at classes, related to this patch -  don't use getters and use the variable itself.

And again, make something like:
```cc
CMemoryPoolManager::GetMemoryPoolMgr()->Destroy(m_factory->mp);
GPOS_DELETE(m_factory);
m_factory = NULL
```

RekGRpth replied:
> Does we really need to use a getter here (`GetFactory`)?

No, but it reduces the diff.

RekGRpth replied:
> make something like:
> 
> ```c++
> CMemoryPoolManager::GetMemoryPoolMgr()->Destroy(m_factory->mp);
> GPOS_DELETE(m_factory);
> m_factory = NULL
> ```

That's exactly what you can't do! First you need to delete, and then release the pool.

HustonMmmavr commented on src/backend/gporca/libgpopt/src/xforms/CXformFactory.cpp:
```diff
@@ -405,10 +392,10 @@ CXformFactory::Init()
+	if (NULL == m_pxff)
+		return;
 
+	CXformFactory *pxff = CXformFactory::Pxff();
```
the usage of m_pxff seems more readable instead of ::Pxff()

RekGRpth replied:
> the usage of m_pxff seems more readable instead of ::Pxff()

Yes, but here it is done to reduce the diff.

HustonMmmavr commented on src/backend/gporca/libgpos/src/error/CMessageRepository.cpp:
```diff
@@ -153,7 +151,10 @@ CMessageRepository::GetMessageRepository()
+		return;
+
+	CMemoryPoolManager::GetMemoryPoolMgr()->Destroy(m_repository->m_mp);
 	CMessageRepository::m_repository = NULL;
```
GPOS_DELETE(m_repository)?

RekGRpth replied:
> GPOS_DELETE(m_repository)?

NO!!! The comment before the method reminds of this: https://github.com/arenadata/gpdb/blob/7189fef49afc8296d897c6386ecd5e237b918a39/src/backend/gporca/libgpos/src/error/CMessageRepository.cpp#L148-L150

HustonMmmavr replied:
It means that messages stored at CMessageRepository, which is loaded [here](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/gporca/libgpos/src/error/CMessageRepository.cpp#L257) from the [static array](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/gporca/libgpos/src/error/CMessage.cpp#L147) would be cleaned by the OS loader, am I right? If so, seems it there would be a leak of allocated m_repository and it's fields, or the deletion of mp will release it?

RekGRpth replied:
> messages stored at CMessageRepository, which is loaded [here](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/gporca/libgpos/src/error/CMessageRepository.cpp#L257) from the [static array](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/gporca/libgpos/src/error/CMessage.cpp#L147) would be cleaned by the OS loader, am I right?

I think yes.

RekGRpth replied:
> If so, seems it there would be a leak of allocated m_repository and it's fields, or the deletion of mp will release it?

I was able to add the delete.

HustonMmmavr replied:
I've tried to detect is the GPOS_DELETE needed here locally by debugging the next code 
```c
// gpos_init
...
CMessageRepository::Init();
CMessageRepository::Shutdown(); 
```

And it's interesting - seems there is no leak of memory on the gpdb which works with the `CMemoryPoolTracker` (which may be set with `optimizer_use_gpdb_allocators` GUC) - the references to memory poll are stored inside each header of allocated memory (so `GPOS_DELETE` called without `mp` argument finds related memory poo without any probleml), and the memory pool, itself, maintains a list of it's allocation. So the call 	`CMemoryPoolManager::Destroy(m_repository->m_mp);` would destroy all allocations. But I don't know the behavior of `CMemoryPoolPalloc`.

RekGRpth replied:
> GPOS_DELETE(m_repository)?

added

HustonMmmavr commented on src/backend/gporca/libgpopt/src/xforms/CXformFactory.cpp:
```diff
@@ -353,13 +353,11 @@ CXformFactory::IsXformIdUsed(CXform::EXformId exfid)
 	GPOS_ASSERT(NULL == Pxff() && "Xform factory was already initialized");
 
-	GPOS_RESULT eres = GPOS_OK;
-
```
I think mixing the Pxff() method (which returns pointer to static m_pxff) and acessing m_pxff explicitly is misleading, 
so assert above may be rewritten like `GPOS_ASSERT(NULL == m_pxff ...)`
I think the ctors of all classes, related to this patch should be refactored and use the unified approach like it's done at CMessageRepository - there is catching and rethrowing of exception (it seems redundant) , instead of it used CAutoMemorryPool with further detach, so seems it may be rewritten like:

```cc
	GPOS_ASSERT(NULL == m_pxff ...);

	CAutoMemoryPool amp;
	CMemoryPool *mp = amp.Pmp();

	m_pxff = GPOS_NEW(mp) CXformFactory(mp);

	// detach safety
	(void) amp.Detach();
```

RekGRpth replied:
> I think the ctors of all classes, related to this patch should be refactored and use the unified approach like it's done at CMessageRepository - there is catching and rethrowing of exception (it seems redundant) , instead of it used CAutoMemorryPool with further detach, so seems it may be rewritten like:

But I think that we need to get rid of CAutoMemoryPool during initialization.

RekGRpth replied:
> I think mixing the Pxff() method (which returns pointer to static m_pxff) and acessing m_pxff explicitly is misleading,
> so assert above may be rewritten like `GPOS_ASSERT(NULL == m_pxff ...)`

Yes, but here it is done to reduce the diff.

RekGRpth replied:
> GPOS_ASSERT(NULL == m_pxff ...)

done

HustonMmmavr commented on src/backend/gporca/libgpos/src/task/CWorkerPoolManager.cpp:
```diff
@@ -80,16 +81,9 @@ CWorkerPoolManager::Init()
 
 		CWorkerPoolManager::m_worker_pool_manager = NULL;
 
-		if (GPOS_MATCH_EX(ex, CException::ExmaSystem, CException::ExmiOOM))
```
Use the same approach as at CMessageRepository?

RekGRpth replied:
> Use the same approach as at CMessageRepository?

No, I think that we need to get rid of CAutoMemoryPool during initialization.

HustonMmmavr commented:
The changes is ok, but seems it's worth to make a bit refacotor of ORCA code to unify approaches

andr-sokolov commented on src/backend/gporca/libgpos/src/_api.cpp:
```diff
@@ -131,41 +131,41 @@ gpos_init(struct gpos_init_params *params)
 	{
-		return;
-	}
+		gpos::CMemoryPoolManager::Init();
```
The gpos::CMemoryPoolManager::Init() method is called here, but below the corresponding Shutdown() is called as CMemoryPoolManager::Shutdown(), i.e. without namespace specification. Now it is ok, but if someday a new CMemoryPoolManager class  appear in the default namespace, then the Shutdown() from another class would be called. I suggest to call Init() without namespace, because of "using namespace gpos;"

RekGRpth replied:
done

andr-sokolov commented on src/backend/gporca/libgpos/src/_api.cpp:
```diff
@@ -131,41 +131,41 @@ gpos_init(struct gpos_init_params *params)
-		return;
-	}
+#ifdef GPOS_DEBUG_COUNTERS
+		CDebugCounter::Shutdown();
```
The gpos_terminate() function is very similar to this catch block. Is it possible to rewrite gpos_terminate() and call it here?

RekGRpth replied:
> Is it possible to rewrite gpos_terminate() and call it here?

gpos_terminate does nothing on release builds, but here I still release the allocated memory in case of errors.

RekGRpth replied:
Although, I did it anyway.

RekGRpth commented:
This patch will be simplified after [patch](https://github.com/arenadata/gpdb/pull/555) work.

HustonMmmavr commented on src/backend/gporca/libgpopt/src/xforms/CXformFactory.cpp:
```diff
@@ -374,23 +372,12 @@ CXformFactory::Init()
 		CMemoryPoolManager::GetMemoryPoolMgr()->Destroy(mp);
 		m_pxff = NULL;
```
does the nulling of static variable (and at other ctors too) is needed, does the GPOS_NEW would change this variable in case of exception?

RekGRpth replied:
https://github.com/arenadata/gpdb/pull/525#issuecomment-1623073108

RekGRpth replied:
> does the nulling of static variable (and at other ctors too) is needed

Yes, this is then checked in the assert in the destructor.

HustonMmmavr replied:
I agree, that destructor checks it for the `NULL`, but seems in case of failed [`Init`](https://github.com/arenadata/gpdb/blob/1fdfa24975800876cc6cdb568e1690409db958b6/src/backend/gporca/libgpopt/src/xforms/CXformFactory.cpp#L366-L372), GPOS_NEW will raise an exception and this variable won't changed (it was a NULL and doesn't changed)

HustonMmmavr replied:
The comment was solved in ADBDEV-3913 - no changes needed https://github.com/arenadata/gpdb/pull/555#discussion_r1257528940

HustonMmmavr commented:
To summarize about `CAutoMemoryPool` (approach with using it inside Init stage may give more clean code), but you are right, this class is not safe - it may throw an exception at it's desctructor, not only at debug build (at debug build it's seems ugly, but not critical), it calls [CMemoryPoolManager::GetMemoryPoolMgr()->Destroy](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/gporca/libgpos/src/memory/CAutoMemoryPool.cpp#L125), it eventually calls `DeleteImpl`, which may raise an exception from [CMemoryPoolPalloc](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/gpopt/utils/CMemoryPoolPalloc.cpp#L63) or [CMemoryPoolTracker](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/gporca/libgpos/src/memory/CMemoryPoolTracker.cpp#L124) (and there CMemoryPoolManager::Destroy may raise at other places)

HustonMmmavr commented on src/backend/gpopt/CGPOptimizer.cpp:
```diff
@@ -211,6 +211,12 @@ extern "C" {
 PlannedStmt *
 GPOPTOptimizedPlan(Query *query, bool *had_unexpected_failure)
 {
+	if (OptimizerMemoryContext == NULL)
```
The next method `SerializeDXLPlan` doesn't need to have the same check?

RekGRpth replied:
fixed

andr-sokolov commented:
@RekGRpth , describe please in the PR description why shutdown argument has been added

RekGRpth commented:
> why shutdown argument has been added

Because initially the shutdown methods were called only on debug builds on exit. And I just kept this behavior.

RekGRpth commented:
> 5\. "worker = worker = ". Is this a typo?

fixed

RekGRpth commented:
> "sequence 7x" - may be "7f"?

Yes, it is really 7x https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/utils/mmgr/aset.c#L740-L747

fixed

RekGRpth commented:
> 2\. The call stack should be given without ... function arguments...

Why?! I may be significant:
```gdb
#4  0x0000000000f647c5 in gpos::CMemoryPoolManager::CreateMemoryPool (this=0x0) at CMemoryPoolManager.cpp:85
```

RekGRpth commented:
> 2\. The call stack should be given without redundant data: without sequence numbers, addresses

done

RekGRpth commented:
> 1. A mechanism for reproducing the error without editing the code should be described

fixed

RekGRpth commented:
> 1. In the second case, it is unclear what kind of real exception we are talking about.

fixed

RekGRpth commented:
> 3\. and where about how GPDB works after the changes

This is said after the words "I refactored the ORCA initialization code"

HustonMmmavr commented on src/backend/gporca/libgpos/src/error/CFSimulator.cpp:
```diff
@@ -131,18 +131,13 @@ CFSimulator::NewStack(ULONG major, ULONG minor)
+	CMemoryPool *mp =
+		CMemoryPoolManager::GetMemoryPoolMgr()->CreateMemoryPool();
 
 	CFSimulator::m_fsim = GPOS_NEW(mp) CFSimulator(mp, GPOS_FSIM_RESOLUTION);
```
Doesn't we need to wrap it into try/catch, like it's done in other `Init` methods and free the `mp` at catch, if error occurred at stage of creation of CFSimulator?

RekGRpth replied:
> if error occurred at stage of creation of CFSimulator

then CFSimulator::Shutdown method is called and it frees mp

HustonMmmavr replied:
but what if Shutdown won't be called for some reasons (for ex. Init will be used in other part of code, but Shutdown won't be by mistake)?
As I said before, seems it'll be nice to unify approaches. Other Init methods do the things with try/catch, and it for me it seems safe or it's a redundant work?

RekGRpth replied:
> Other Init methods do the things with try/catch

They have a condition in shutdown methods for destroy mp, but CFSimulator has not!
If I destroy mp in CFSimulator::Init, how I destroy it in CFSimulator::Shutdown? Nullify mp and then check mp for NULL?

HustonMmmavr replied:
Sorry, i thought that Shutdown of CFSimulator is static function, but it's not. Ok what would be if the exception would occur on this instruction? As I understand `m_fsim` would be `NULL`, TerminateGPOPT would call gpos_terminate and here is a check, which should fail:
```
	if (NULL != CFSimulator::FSim())
	{
		CFSimulator::FSim()->Shutdown();
	}
```
So possible leak
Am I wrong?

RekGRpth replied:
Yes, you are right, I corrected it.

HustonMmmavr replied:
Please, verify other places where AutoMemoryPool was changed

RekGRpth replied:
done

andr-sokolov commented on src/backend/gpopt/CGPOptimizer.cpp:
```diff
@@ -245,14 +257,34 @@ InitGPOPT()
+		{
+			CGPOptimizer::TerminateGPOPT(true);
+		}
+		GPOS_CATCH_EX(ex)
```
I suggest to use different names for the CException variables to easily distinguish them.

RekGRpth replied:
done

andr-sokolov commented on src/backend/gporca/libgpopt/src/init.cpp:
```diff
@@ -67,15 +56,28 @@ gpopt_init()
+		CMDCache::Shutdown();
+	}
 
 	CMemoryPoolManager::GetMemoryPoolMgr()->Destroy(mp);
```
mp may be NULL if malloc at the call stack below returns null.
```
gpos::clib::Malloc (size=304) at clibwrapper.cpp:475
gpos::CMemoryPoolManager::SetupGlobalMemoryPoolManager<gpos::CMemoryPoolPallocManager, gpos::CMemoryPoolPalloc> () at CMemoryPoolManager.h:109
gpos::CMemoryPoolPallocManager::Init () at CMemoryPoolPallocManager.cpp:57
CGPOptimizer::InitGPOPT () at CGPOptimizer.cpp:176
InitGPOPT () at CGPOptimizer.cpp:260
InitPostgres () at postinit.c:664
```
There is GPOS_ASSERT(NULL != mp) at the beginning of CMemoryPoolManager::Destroy. So let's add `if (NULL != mp)` here

RekGRpth replied:
fixed

andr-sokolov commented on src/backend/gporca/libgpos/src/_api.cpp:
```diff
@@ -280,20 +257,47 @@ gpos_exec(gpos_exec_params *params)
-	CMemoryPoolManager::GetMemoryPoolMgr()->Shutdown();
-#endif	// GPOS_DEBUG
+
+	if (NULL != CMessageRepository::GetMessageRepository())
```
CMessageRepository::GetMessageRepository() always returns not NULL, because of `GPOS_ASSERT(NULL != m_repository);` in this method. If malloc returns 0 at `gpos::clib::Malloc (size=304) at clibwrapper.cpp:475`, then m_repository is NULL

RekGRpth replied:
fixed

andr-sokolov commented on src/backend/gpopt/CGPOptimizer.cpp:
```diff
@@ -245,14 +257,34 @@ InitGPOPT()
+		errstart(WARNING, ex.Filename(), ex.Line(), NULL, TEXTDOMAIN);
+		errfinish(errcode(ERRCODE_INTERNAL_ERROR), errprintstack(true));
+
+		if (OptimizerMemoryContext != NULL)
```
This block is never executed, because errfinish() terminates the process.

RekGRpth replied:
> errfinish() terminates the process

Why?! errstart(WARNING...

RekGRpth replied:
> errfinish() terminates the process

No! Process is terminated by debugger when it is attaching to new process at https://github.com/arenadata/gpdb/blob/20537e17098e824f6296366478821ada8e98b67e/src/backend/utils/error/elog.c#L3686 on printing stack trace

andr-sokolov commented on src/backend/gporca/libnaucrates/src/init.cpp:
```diff
@@ -151,9 +141,17 @@ gpdxl_init()
+		return;
+	}
+
 	ShutdownDXL();
```
If malloc returns 0 at `gpos::clib::Malloc (size=304) at clibwrapper.cpp:475`, then pmpXerces is NULL. ShutdownDXL has `GPOS_ASSERT(NULL != pmpXerces)` at the beginning

RekGRpth replied:
fixed

andr-sokolov commented:
@RekGRpth If malloc() in operator new returns NULL, then bad_alloc exception is thrown, because handler is not set. The exception is not caught, so the process is terminated without messages in the log. GPOS_CATCH_EX can catch gpos::CException only. Stack to reproduce the error:
```
operator new (sz=48) at new_op.cc:52
gpos::CMemoryPoolManager::SetupGlobalMemoryPoolManager<gpos::CMemoryPoolPallocManager, gpos::CMemoryPoolPalloc> () at CMemoryPoolManager.h:119
gpos::CMemoryPoolPallocManager::Init () at CMemoryPoolPallocManager.cpp:57
CGPOptimizer::InitGPOPT () at CGPOptimizer.cpp:176
InitGPOPT () at CGPOptimizer.cpp:260
```

RekGRpth commented:
> @RekGRpth If malloc() in operator new returns NULL, then bad_alloc exception is thrown, because handler is not set. The exception is not caught, so the process is terminated without messages in the log. GPOS_CATCH_EX can catch gpos::CException only. Stack to reproduce the error:
> 
> ```
> operator new (sz=48) at new_op.cc:52
> gpos::CMemoryPoolManager::SetupGlobalMemoryPoolManager<gpos::CMemoryPoolPallocManager, gpos::CMemoryPoolPalloc> () at CMemoryPoolManager.h:119
> gpos::CMemoryPoolPallocManager::Init () at CMemoryPoolPallocManager.cpp:57
> CGPOptimizer::InitGPOPT () at CGPOptimizer.cpp:176
> InitGPOPT () at CGPOptimizer.cpp:260
> ```

fixed

andr-sokolov commented on src/backend/gpopt/CGPOptimizer.cpp:
```diff
@@ -272,14 +288,30 @@ TerminateGPOPT()
 {
 	GPOS_TRY
 	{
-		return CGPOptimizer::TerminateGPOPT();
```
Do we really need to rewrite TerminateGPOPT()?

RekGRpth replied:
The TerminateGPOPT function has been rewritten to match the InitGPOPT function.

RekGRpth replied:
reverted

andr-sokolov commented on src/backend/gporca/libgpopt/src/init.cpp:
```diff
@@ -41,21 +40,11 @@ static CMemoryPool *mp = NULL;
-	{
-		CAutoMemoryPool amp;
-		mp = amp.Pmp();
+	mp = CMemoryPoolManager::GetMemoryPoolMgr()->CreateMemoryPool();
```
Do this change and the same changes below relate to OOM?

RekGRpth replied:
Yes, this was done to avoid calling the auto memory pool's leaky destructor with an exception.

andr-sokolov commented on src/backend/gporca/libnaucrates/src/init.cpp:
```diff
@@ -122,23 +120,13 @@ void
 gpdxl_init()
 {
 	// create memory pool for Xerces global allocations
-	{
```
Add in the description why the gpdxl_init() function has been modified. This changes do not fix errors at processing of OOM and exceptions

RekGRpth replied:
The gpdxl_init function has been changed, because the auto memory pool is used to automatically free memory when it goes out of scope, but here it is immediately removed from it, and in fact, it is not needed at all.

andr-sokolov commented:
@RekGRpth , Rewrite please description to use past tense to tell about the error, present tense to tell about right behaviour. It will be easier to understand what the patch changes and what does not.

RekGRpth commented:
> use past tense to tell about the error, present tense to tell about right behaviour

done

andr-sokolov commented:
@RekGRpth , I suggest to replace 2nd and 3rd points with something like "ORCA does not work correctly when initialization is not fully completed". It is obvious and will be enough to explain why backend is terminated on exception after the patch.

RekGRpth commented:
> replace 2nd and 3rd points with something like "ORCA does not work correctly when initialization is not fully completed"

done (in commit message)

andr-sokolov commented:
@RekGRpth My version of commit message
```
ORCA initialization refactoring

When initializing ORCA mixed a C approach (returning error codes and then
checking them) and a C++ approach (throwing exceptions and then catching them).
This mixing led to errors. For example, in case of low memory or any other
exception, the CWorkerPoolManager::Init function did not return GPOS_OK, and
then the gpos_init function called
the CMemoryPoolManager::GetMemoryPoolMgr()->Shutdown function, which reset
CMemoryPoolManager::m_memory_pool_mgr. In this case, all the exceptions that
arose were not rethrown, so the execution continued, and the following
gpdxl_init function was called, in which the CAutoMemoryPool class was
initialized. In the constructor of this class, the static method
CMemoryPoolManager::GetMemoryPoolMgr returned NULL, because we nullified
m_memory_pool_mgr up the stack. Thus, a segfault could occur.

The patch replaces C approach with C++ approach in ORCA initialization and adds
processing of standard C++ exceptions, which are thrown for example by the new
operator in case of out of memory.

Now a backend terminates with error on any exception during ORCA initialization,
because ORCA does not work correctly when initialization is not fully completed.

The gpdxl_init function has been changed, because the auto memory pool is used
to automatically free memory when it goes out of scope, but here the pools were
immediately removed, and in fact, they are not needed here at all.
```

RekGRpth commented:
> @RekGRpth My version of commit message
> 
> ```
> ORCA initialization refactoring
> 
> When initializing ORCA mixed a C approach (returning error codes and then
> checking them) and a C++ approach (throwing exceptions and then catching them).
> This mixing led to errors. For example, in case of low memory or any other
> exception, the CWorkerPoolManager::Init function did not return GPOS_OK, and
> then the gpos_init function called
> the CMemoryPoolManager::GetMemoryPoolMgr()->Shutdown function, which reset
> CMemoryPoolManager::m_memory_pool_mgr. In this case, all the exceptions that
> arose were not rethrown, so the execution continued, and the following
> gpdxl_init function was called, in which the CAutoMemoryPool class was
> initialized. In the constructor of this class, the static method
> CMemoryPoolManager::GetMemoryPoolMgr returned NULL, because we nullified
> m_memory_pool_mgr up the stack. Thus, a segfault could occur.
> 
> The patch replaces C approach with C++ approach in ORCA initialization and adds
> processing of standard C++ exceptions, which are thrown for example by the new
> operator in case of out of memory.
> 
> Now a backend terminates with error on any exception during ORCA initialization,
> because ORCA does not work correctly when initialization is not fully completed.
> 
> The gpdxl_init function has been changed, because the auto memory pool is used
> to automatically free memory when it goes out of scope, but here the pools were
> immediately removed, and in fact, they are not needed here at all.
> ```

updated

RekGRpth committed:
ORCA initialization refactoring

When initializing ORCA mixed a C approach (returning error codes and then
checking them) and a C++ approach (throwing exceptions and then catching them).
This mixing led to errors. For example, in case of low memory or any other
exception, the CWorkerPoolManager::Init function did not return GPOS_OK, and
then the gpos_init function called
the CMemoryPoolManager::GetMemoryPoolMgr()->Shutdown function, which reset
CMemoryPoolManager::m_memory_pool_mgr. In this case, all the exceptions that
arose were not rethrown, so the execution continued, and the following
gpdxl_init function was called, in which the CAutoMemoryPool class was
initialized. In the constructor of this class, the static method
CMemoryPoolManager::GetMemoryPoolMgr returned NULL, because we nullified
m_memory_pool_mgr up the stack. Thus, a segfault could occur.

The patch replaces C approach with C++ approach in ORCA initialization and adds
processing of standard C++ exceptions, which are thrown for example by the new
operator in case of out of memory.

Now a backend terminates with error on any exception during ORCA initialization,
because ORCA does not work correctly when initialization is not fully completed.

CAutoMemoryPool has been replaced with this class constructor content in the
places, where CAutoMemoryPool destructor could be called after an exception. In
this cases we could get exception in CAutoMemoryPool destructor during
processing of other exception. This caused to backend termination without
informative message.

The gpdxl_init function has been changed, because the auto memory pool is used
to automatically free memory when it goes out of scope, but here the pools were
immediately removed, and in fact, they are not needed here at all.

Stolb27 commented:
Only known flaky was met. I'm going to rerun tests after merging interrupts holdoff fix

Stolb27 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-3674

RekGRpth commented:
It seems we missed the upstream!
Looks like they peeped the implementation here!
https://github.com/greenplum-db/gpdb/pull/16085
